File: dashboard/src/services/manualImport.js
// 手動匯入解析器：與原本後端 manual_import.py 行為對齊，
// HTML 解析改用瀏覽器內建 DOMParser。

const JOB_URL_PATTERN = /(?:https?:)?\/\/www\.104\.com\.tw\/job\/([a-z0-9]+)/i
const CUSTNAME_PATTERN = /custname="([^"]+)"/
const MAPS_PATTERN = /https:\/\/maps\.google\.com\.tw\/\?q=([^@"&]+)/

function normalizeText(value) {
  if (!value) {
    return ''
  }
  return String(value).replace(/\s+/g, ' ').trim()
}

function normalizeUrl(rawUrl) {
  const url = String(rawUrl || '')
  if (url.startsWith('//')) {
    return `https:${url}`
  }
  if (url.startsWith('/')) {
    return new URL(url, 'https://www.104.com.tw').href
  }
  return url
}

function normalizeMultiline(value) {
  if (!value) {
    return ''
  }
  return String(value)
    .split(/\r?\n/)
    .map(normalizeText)
    .filter(Boolean)
    .join('\n')
}

function unescapeHtml(value) {
  const doc = new DOMParser().parseFromString(String(value), 'text/html')
  return doc.documentElement.textContent || ''
}

async function stableManualId(value) {
  const data = new TextEncoder().encode(value)
  const digest = await crypto.subtle.digest('SHA-1', data)
  const hex = [...new Uint8Array(digest)].map(byte => byte.toString(16).padStart(2, '0')).join('')
  return `manual-${hex.slice(0, 12)}`
}

// 對齊 Python 版以 f-string 內插 list 產生的字串（tags 參與 job_id 雜湊時的格式）
function pyListRepr(items) {
  return `[${items.map(item => `'${item}'`).join(', ')}]`
}

function scrapedAtNow() {
  return new Date().toISOString().replace('Z', '+00:00')
}

function makeJob(fields) {
  return {
    job_id: '',
    source: '',
    keyword: '',
    title: '',
    company_name: '',
    location: '',
    salary: '',
    experience: '',
    education: '',
    description: '',
    requirement: '',
    tags: '',
    job_url: '',
    scraped_at: scrapedAtNow(),
    ...fields
  }
}

function inferCompanyName(content) {
  const custMatch = CUSTNAME_PATTERN.exec(content)
  if (custMatch) {
    return normalizeText(unescapeHtml(custMatch[1]))
  }
  const mapsMatch = MAPS_PATTERN.exec(content)
  if (mapsMatch) {
    try {
      return normalizeText(decodeURIComponent(mapsMatch[1]))
    } catch {
      return normalizeText(mapsMatch[1])
    }
  }
  return ''
}

function resolveImportContext(content, keyword, companyName) {
  const resolvedCompany = normalizeText(companyName) || inferCompanyName(content)
  let resolvedKeyword = normalizeText(keyword)
  if (!resolvedKeyword) {
    resolvedKeyword = `${resolvedCompany || '104'} 手動匯入`
  }
  return { keyword: resolvedKeyword, companyName: resolvedCompany }
}

// 對齊 Python HTMLParser 逐段收集文字再以空白串接的行為
function textJoin(element) {
  const walker = element.ownerDocument.createTreeWalker(element, NodeFilter.SHOW_TEXT)
  const chunks = []
  while (walker.nextNode()) {
    const text = normalizeText(walker.currentNode.nodeValue)
    if (text) {
      chunks.push(text)
    }
  }
  return normalizeText(chunks.join(' '))
}

function lastNonEmpty(values) {
  for (let i = values.length - 1; i >= 0; i -= 1) {
    if (values[i]) {
      return values[i]
    }
  }
  return ''
}

async function parseListHtmlJobs(doc, { keyword, companyName }) {
  const jobs = []
  const containers = doc.querySelectorAll('div.info-wrapper.job-list-container--cprofile')

  for (const container of containers) {
    const title = lastNonEmpty([...container.querySelectorAll('.info-name')].map(textJoin))
    if (!title) {
      continue
    }

    const anchors = container.querySelectorAll('a[data-gtm-job="點擊工作"]')
    const href = anchors.length ? normalizeUrl(anchors[anchors.length - 1].getAttribute('href') || '') : ''
    const tags = [...container.querySelectorAll('.info-tags__text')].map(textJoin).filter(Boolean)
    const salary =
      lastNonEmpty([...container.querySelectorAll('.info-othertags__text')].map(textJoin)) ||
      normalizeText(container.getAttribute('jobsalarydesc') || '')

    const idMatch = JOB_URL_PATTERN.exec(href)
    const jobId = idMatch ? idMatch[1] : await stableManualId(`${title}|${href}|${pyListRepr(tags)}`)

    jobs.push(
      makeJob({
        job_id: jobId,
        source: '104-manual-html',
        keyword,
        title,
        company_name: companyName,
        location: tags[0] || '',
        experience: tags[1] || '',
        education: tags[2] || '',
        salary,
        tags: tags.join('，'),
        job_url: href
      })
    )
  }
  return jobs
}

function extractDetailTitle(doc) {
  const candidates = [
    doc.querySelector('meta[property="og:title"]')?.getAttribute('content'),
    doc.querySelector('meta[name="title"]')?.getAttribute('content'),
    doc.querySelector('h1')?.textContent,
    doc.querySelector('[class*="job-title"], [class*="jobName"], [class*="job-name"]')?.textContent
  ]
  for (const candidate of candidates) {
    const title = normalizeText(candidate)
    if (title) {
      return title
    }
  }
  return ''
}

function extractDetailDescription(doc) {
  const attrDescription = doc.querySelector('[jobdescription]')?.getAttribute('jobdescription')
  if (attrDescription) {
    return normalizeMultiline(attrDescription)
  }
  const element = doc.querySelector('[class*="job-description__content"]')
  return element ? normalizeMultiline(element.textContent) : ''
}

function splitDescriptionRequirement(description) {
  const marker = '【需求條件】'
  if (!description.includes(marker)) {
    return { description, requirement: '' }
  }
  const index = description.indexOf(marker)
  return {
    description: normalizeMultiline(description.slice(0, index)),
    requirement: normalizeMultiline(description.slice(index + marker.length))
  }
}

function extractDetailCategories(doc) {
  const categories = []
  doc.querySelectorAll('[data-gtm-content="職務類別"] u').forEach(node => {
    const category = normalizeText(node.textContent)
    if (category && !categories.includes(category)) {
      categories.push(category)
    }
  })
  return categories
}

function extractDetailLocation(doc) {
  const area = normalizeText(doc.querySelector('[addressarea]')?.getAttribute('addressarea'))
  const address = normalizeText(doc.querySelector('.job-address span')?.textContent)
  if (area && address && !address.startsWith(area)) {
    return `${area} ${address}`
  }
  return address || area
}

async function parseDetailHtmlJob(doc, rawContent, { keyword, companyName }) {
  const rawDescription = extractDetailDescription(doc)
  const salary = normalizeText(doc.querySelector('[salary]')?.getAttribute('salary'))
  const location = extractDetailLocation(doc)
  const categories = extractDetailCategories(doc)

  if (!rawDescription && !salary && !location && !categories.length) {
    return []
  }

  const title = extractDetailTitle(doc) || '單一職缺詳細內容'
  const { description, requirement } = splitDescriptionRequirement(rawDescription)
  const urlMatch = /(?:https?:)?\/\/www\.104\.com\.tw\/job\/[a-z0-9]+[^"\s<]*/i.exec(rawContent)
  const jobUrl = urlMatch ? normalizeUrl(urlMatch[0]) : ''
  const idMatch = JOB_URL_PATTERN.exec(jobUrl)
  const jobId = idMatch
    ? idMatch[1]
    : await stableManualId(`${title}|${companyName}|${location}|${description.slice(0, 80)}`)

  return [
    makeJob({
      job_id: jobId,
      source: 'manual-html-detail',
      keyword,
      title,
      company_name: companyName,
      location,
      salary,
      description,
      requirement,
      tags: categories.join('，'),
      job_url: jobUrl
    })
  ]
}

async function parseHtmlJobs(content, context) {
  const doc = new DOMParser().parseFromString(content, 'text/html')
  const listJobs = await parseListHtmlJobs(doc, context)
  if (listJobs.length) {
    return listJobs
  }
  return parseDetailHtmlJob(doc, content, context)
}

// 極簡 RFC 4180 CSV 解析：支援引號欄位、跳脫的雙引號與欄內換行
function parseCsvRows(text) {
  const rows = []
  let row = []
  let field = ''
  let inQuotes = false

  for (let i = 0; i < text.length; i += 1) {
    const char = text[i]
    if (inQuotes) {
      if (char === '"') {
        if (text[i + 1] === '"') {
          field += '"'
          i += 1
        } else {
          inQuotes = false
        }
      } else {
        field += char
      }
    } else if (char === '"') {
      inQuotes = true
    } else if (char === ',') {
      row.push(field)
      field = ''
    } else if (char === '\n' || char === '\r') {
      if (char === '\r' && text[i + 1] === '\n') {
        i += 1
      }
      row.push(field)
      field = ''
      rows.push(row)
      row = []
    } else {
      field += char
    }
  }
  if (field !== '' || row.length) {
    row.push(field)
    rows.push(row)
  }
  return rows.filter(cells => cells.some(cell => cell.trim() !== ''))
}

async function parseCsvJobs(content, { keyword, companyName }) {
  const rows = parseCsvRows(content)
  if (rows.length < 2) {
    return []
  }
  const headers = rows[0].map(header => header.trim())
  const jobs = []

  for (const cells of rows.slice(1)) {
    const record = {}
    headers.forEach((header, index) => {
      record[header] = cells[index] ?? ''
    })
    const pick = (...keys) => {
      for (const key of keys) {
        if (record[key]) {
          return record[key]
        }
      }
      return ''
    }

    const title = normalizeText(pick('title', '職缺', '職稱'))
    if (!title) {
      continue
    }
    const jobUrl = normalizeText(pick('job_url', 'url', '網址'))
    const idMatch = JOB_URL_PATTERN.exec(jobUrl)
    const jobId = record.job_id || (idMatch ? idMatch[1] : await stableManualId(`${title}|${jobUrl}`))

    jobs.push(
      makeJob({
        job_id: normalizeText(jobId),
        source: 'manual-csv',
        keyword: normalizeText(pick('keyword')) || keyword,
        title,
        company_name: normalizeText(pick('company_name', '公司')) || companyName,
        location: normalizeText(pick('location', '地點')),
        salary: normalizeText(pick('salary', '薪資')),
        experience: normalizeText(pick('experience', '經驗')),
        education: normalizeText(pick('education', '學歷')),
        description: normalizeText(pick('description', '描述')),
        requirement: normalizeText(pick('requirement', '條件')),
        tags: normalizeText(pick('tags', '標籤')),
        job_url: jobUrl
      })
    )
  }
  return jobs
}

async function parseTextJobs(content, { keyword, companyName }) {
  const lines = content
    .split(/\r?\n/)
    .map(normalizeText)
    .filter(Boolean)
  const jobs = []

  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index]
    if (!['工程師', '專員', '經理', '顧問'].some(word => line.includes(word))) {
      continue
    }
    if (line.length > 80) {
      continue
    }
    const nextLine = lines[index + 1] || ''
    const parts = nextLine
      .split(/[｜|]/)
      .map(normalizeText)
      .filter(Boolean)

    jobs.push(
      makeJob({
        job_id: await stableManualId(`${line}|${nextLine}|${index}`),
        source: 'manual-text',
        keyword,
        title: line,
        company_name: companyName,
        location: parts[0] || '',
        experience: parts[1] || '',
        education: parts[2] || '',
        tags: nextLine
      })
    )
  }
  return jobs
}

function detectFormat(text) {
  if (text.includes('<') && text.includes('>')) {
    return 'html'
  }
  const firstLine = text.split(/\r?\n/, 1)[0]
  if (firstLine.includes(',') && (firstLine.includes('title') || firstLine.includes('職缺'))) {
    return 'csv'
  }
  return 'text'
}

export async function parseManualJobs(content, { keyword = '', companyName = '', inputFormat = 'auto' } = {}) {
  const text = String(content || '').trim()
  if (!text) {
    return { inputFormat: 'empty', keyword: '', companyName: '', jobs: [] }
  }

  const context = resolveImportContext(text, keyword, companyName)
  const selectedFormat = inputFormat === 'auto' ? detectFormat(text) : inputFormat

  let jobs
  if (selectedFormat === 'html') {
    jobs = await parseHtmlJobs(text, context)
  } else if (selectedFormat === 'csv') {
    jobs = await parseCsvJobs(text, context)
  } else {
    jobs = await parseTextJobs(text, context)
  }

  return { inputFormat: selectedFormat, keyword: context.keyword, companyName: context.companyName, jobs }
}

File: src/job_collector/manual_import.py
from __future__ import annotations

import csv
import hashlib
import html
import re
from html.parser import HTMLParser
from io import StringIO
from typing import Any
from urllib.parse import unquote

from job_collector.crawler_104 import normalize_text, normalize_url
from job_collector.models import JobPost


JOB_URL_PATTERN = re.compile(r"(?:https?:)?//www\.104\.com\.tw/job/([a-z0-9]+)", re.IGNORECASE)
COMPANY_NAME_PATTERN = re.compile(r'custname="([^"]+)"')
HTML_ATTR_PATTERN_TEMPLATE = r'\b{attr}="([^"]*)"'


def stable_manual_id(value: str) -> str:
    digest = hashlib.sha1(value.encode("utf-8")).hexdigest()[:12]
    return f"manual-{digest}"


def infer_company_name(content: str) -> str:
    match = COMPANY_NAME_PATTERN.search(content)
    if match:
        return normalize_text(html.unescape(match.group(1)))
    maps_match = re.search(r"https://maps\.google\.com\.tw/\?q=([^@\"&]+)", content)
    if maps_match:
        return normalize_text(unquote(maps_match.group(1)))
    return ""


def resolve_import_context(content: str, *, keyword: str, company_name: str) -> tuple[str, str]:
    resolved_company = normalize_text(company_name) or infer_company_name(content)
    resolved_keyword = normalize_text(keyword)
    if not resolved_keyword:
        resolved_keyword = f"{resolved_company or '104'} 手動匯入"
    return resolved_keyword, resolved_company


class CompanyJobHTMLParser(HTMLParser):
    def __init__(self, *, keyword: str, company_name: str) -> None:
        super().__init__(convert_charrefs=True)
        self.keyword = keyword
        self.company_name = company_name
        self.jobs: list[JobPost] = []
        self.current: dict[str, Any] | None = None
        self.capture_stack: list[dict[str, Any]] = []

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        attr = {key: value or "" for key, value in attrs}
        class_names = set(attr.get("class", "").split())

        if self.current:
            self.current["depth"] += 1
            for capture in self.capture_stack:
                capture["depth"] += 1

        if tag == "div" and {"info-wrapper", "job-list-container--cprofile"}.issubset(class_names):
            self.current = {
                "depth": 1,
                "title": "",
                "href": "",
                "tags": [],
                "salary": attr.get("jobsalarydesc", ""),
            }
            return

        if not self.current:
            return

        if tag == "a" and attr.get("data-gtm-job") == "點擊工作":
            self.current["href"] = normalize_url(attr.get("href", ""))

        if "info-name" in class_names:
            self.capture_stack.append({"kind": "title", "depth": 1, "text": []})
        elif "info-tags__text" in class_names:
            self.capture_stack.append({"kind": "tag", "depth": 1, "text": []})
        elif "info-othertags__text" in class_names:
            self.capture_stack.append({"kind": "salary", "depth": 1, "text": []})

    def handle_data(self, data: str) -> None:
        if not self.current:
            return
        text = normalize_text(data)
        if not text:
            return
        for capture in self.capture_stack:
            capture["text"].append(text)

    def handle_endtag(self, tag: str) -> None:
        if not self.current:
            return

        finished: list[dict[str, Any]] = []
        for capture in self.capture_stack:
            capture["depth"] -= 1
            if capture["depth"] <= 0:
                finished.append(capture)

        for capture in finished:
            self.capture_stack.remove(capture)
            text = normalize_text(" ".join(capture["text"]))
            if capture["kind"] == "title" and text:
                self.current["title"] = text
            elif capture["kind"] == "tag" and text:
                self.current["tags"].append(text)
            elif capture["kind"] == "salary" and text:
                self.current["salary"] = text

        self.current["depth"] -= 1
        if self.current["depth"] <= 0:
            self.finish_current_job()

    def finish_current_job(self) -> None:
        if not self.current:
            return

        title = normalize_text(self.current.get("title", ""))
        href = normalize_url(self.current.get("href", ""))
        if not title:
            self.current = None
            return

        tags = [normalize_text(tag) for tag in self.current.get("tags", []) if normalize_text(tag)]
        job_id_match = JOB_URL_PATTERN.search(href)
        job_id = job_id_match.group(1) if job_id_match else stable_manual_id(f"{title}|{href}|{tags}")

        self.jobs.append(
            JobPost(
                job_id=job_id,
                source="104-manual-html",
                keyword=self.keyword,
                title=title,
                company_name=self.company_name,
                location=tags[0] if len(tags) > 0 else "",
                experience=tags[1] if len(tags) > 1 else "",
                education=tags[2] if len(tags) > 2 else "",
                salary=normalize_text(self.current.get("salary", "")),
                tags="，".join(tags),
                job_url=href,
            )
        )
        self.current = None


def parse_html_jobs(content: str, *, keyword: str, company_name: str) -> list[JobPost]:
    parser = CompanyJobHTMLParser(keyword=keyword, company_name=company_name)
    parser.feed(content)
    if parser.jobs:
        return parser.jobs
    return parse_detail_html_job(content, keyword=keyword, company_name=company_name)


def normalize_multiline(value: str | None) -> str:
    if not value:
        return ""
    lines = [normalize_text(line) for line in value.splitlines()]
    return "\n".join(line for line in lines if line)


def extract_html_attr(content: str, attr: str) -> str:
    pattern = HTML_ATTR_PATTERN_TEMPLATE.format(attr=re.escape(attr))
    match = re.search(pattern, content, flags=re.IGNORECASE | re.DOTALL)
    if not match:
        return ""
    return html.unescape(match.group(1))


def strip_tags(value: str) -> str:
    return html.unescape(re.sub(r"<[^>]+>", " ", value))


def extract_detail_title(content: str) -> str:
    patterns = [
        r'<meta[^>]+property="og:title"[^>]+content="([^"]+)"',
        r'<meta[^>]+name="title"[^>]+content="([^"]+)"',
        r'<h1[^>]*>(.*?)</h1>',
        r'class="[^"]*(?:job-title|jobName|job-name)[^"]*"[^>]*>(.*?)<',
    ]
    for pattern in patterns:
        match = re.search(pattern, content, flags=re.IGNORECASE | re.DOTALL)
        if match:
            title = normalize_text(strip_tags(match.group(1)))
            if title:
                return title
    return ""


def extract_detail_description(content: str) -> str:
    attr_description = extract_html_attr(content, "jobdescription")
    if attr_description:
        return normalize_multiline(attr_description)

    match = re.search(
        r'class="[^"]*job-description__content[^"]*"[^>]*>(.*?)</p>',
        content,
        flags=re.IGNORECASE | re.DOTALL,
    )
    if match:
        return normalize_multiline(strip_tags(match.group(1)))
    return ""


def split_description_requirement(description: str) -> tuple[str, str]:
    marker = "【需求條件】"
    if marker not in description:
        return description, ""
    before, after = description.split(marker, 1)
    return normalize_multiline(before), normalize_multiline(after)


def extract_detail_categories(content: str) -> list[str]:
    categories = []
    for match in re.finditer(
        r'data-gtm-content="職務類別".*?<u[^>]*>(.*?)</u>',
        content,
        flags=re.IGNORECASE | re.DOTALL,
    ):
        category = normalize_text(strip_tags(match.group(1)))
        if category and category not in categories:
            categories.append(category)
    return categories


def extract_detail_location(content: str) -> str:
    area = normalize_text(extract_html_attr(content, "addressarea"))
    address_match = re.search(
        r'class="[^"]*job-address[^"]*"[^>]*>.*?<span[^>]*>(.*?)</span>',
        content,
        flags=re.IGNORECASE | re.DOTALL,
    )
    address = normalize_text(strip_tags(address_match.group(1))) if address_match else ""
    if area and address and not address.startswith(area):
        return f"{area} {address}"
    return address or area


def parse_detail_html_job(content: str, *, keyword: str, company_name: str) -> list[JobPost]:
    description = extract_detail_description(content)
    salary = normalize_text(extract_html_attr(content, "salary"))
    location = extract_detail_location(content)
    categories = extract_detail_categories(content)

    if not any([description, salary, location, categories]):
        return []

    title = extract_detail_title(content) or "單一職缺詳細內容"
    description, requirement = split_description_requirement(description)
    tags = "，".join(categories)
    job_url_match = re.search(r'(?:https?:)?//www\.104\.com\.tw/job/[a-z0-9]+[^"\s<]*', content, flags=re.IGNORECASE)
    job_url = normalize_url(job_url_match.group(0)) if job_url_match else ""
    job_id_match = JOB_URL_PATTERN.search(job_url)
    job_id = job_id_match.group(1) if job_id_match else stable_manual_id(f"{title}|{company_name}|{location}|{description[:80]}")

    return [
        JobPost(
            job_id=job_id,
            source="manual-html-detail",
            keyword=keyword,
            title=title,
            company_name=company_name,
            location=location,
            salary=salary,
            description=description,
            requirement=requirement,
            tags=tags,
            job_url=job_url,
        )
    ]


def parse_csv_jobs(content: str, *, keyword: str, company_name: str) -> list[JobPost]:
    reader = csv.DictReader(StringIO(content))
    jobs: list[JobPost] = []

    for row in reader:
        title = normalize_text(row.get("title") or row.get("職缺") or row.get("職稱") or "")
        if not title:
            continue
        job_url = normalize_text(row.get("job_url") or row.get("url") or row.get("網址") or "")
        match = JOB_URL_PATTERN.search(job_url)
        job_id = row.get("job_id") or (match.group(1) if match else stable_manual_id(f"{title}|{job_url}"))
        jobs.append(
            JobPost(
                job_id=normalize_text(job_id),
                source="manual-csv",
                keyword=normalize_text(row.get("keyword") or keyword),
                title=title,
                company_name=normalize_text(row.get("company_name") or row.get("公司") or company_name),
                location=normalize_text(row.get("location") or row.get("地點") or ""),
                salary=normalize_text(row.get("salary") or row.get("薪資") or ""),
                experience=normalize_text(row.get("experience") or row.get("經驗") or ""),
                education=normalize_text(row.get("education") or row.get("學歷") or ""),
                description=normalize_text(row.get("description") or row.get("描述") or ""),
                requirement=normalize_text(row.get("requirement") or row.get("條件") or ""),
                tags=normalize_text(row.get("tags") or row.get("標籤") or ""),
                job_url=job_url,
            )
        )
    return jobs


def parse_text_jobs(content: str, *, keyword: str, company_name: str) -> list[JobPost]:
    lines = [normalize_text(line) for line in content.splitlines()]
    lines = [line for line in lines if line]
    jobs: list[JobPost] = []

    for index, line in enumerate(lines):
        if "工程師" not in line and "專員" not in line and "經理" not in line and "顧問" not in line:
            continue
        if len(line) > 80:
            continue
        next_line = lines[index + 1] if index + 1 < len(lines) else ""
        parts = [normalize_text(part) for part in re.split(r"[｜|]", next_line) if normalize_text(part)]
        jobs.append(
            JobPost(
                job_id=stable_manual_id(f"{line}|{next_line}|{index}"),
                source="manual-text",
                keyword=keyword,
                title=line,
                company_name=company_name,
                location=parts[0] if len(parts) > 0 else "",
                experience=parts[1] if len(parts) > 1 else "",
                education=parts[2] if len(parts) > 2 else "",
                tags=next_line,
            )
        )
    return jobs


def parse_manual_jobs(
    content: str,
    *,
    keyword: str,
    company_name: str,
    input_format: str = "auto",
) -> tuple[str, list[JobPost]]:
    text = content.strip()
    if not text:
        return "empty", []

    keyword, company_name = resolve_import_context(text, keyword=keyword, company_name=company_name)

    selected_format = input_format
    if input_format == "auto":
        if "<" in text and ">" in text:
            selected_format = "html"
        elif "," in text.splitlines()[0] and ("title" in text.splitlines()[0] or "職缺" in text.splitlines()[0]):
            selected_format = "csv"
        else:
            selected_format = "text"

    if selected_format == "html":
        return selected_format, parse_html_jobs(text, keyword=keyword, company_name=company_name)
    if selected_format == "csv":
        return selected_format, parse_csv_jobs(text, keyword=keyword, company_name=company_name)
    return selected_format, parse_text_jobs(text, keyword=keyword, company_name=company_name)
